### ISSUE TITLE
Buildkite CI: Solve non-core member CI running security issue

### ISSUE THREAD

**Issue by bkase** [CodaProtocol/coda#4988]
We want Buildkite jobs to appear on PRs in the same way that circle jobs appear on PRs automatically.

For security reasons, we don't want to run arbitrary code in a PR from someone who is not on the core team. We need to make sure to detect this.

See a comment below for the latest proposal.

Historical proposals:

Proposals:
1. Hit GitHub API, if PR is from a fork; deploy a small github bot that listens to a special comment from a core team member (example "!start-test") then allow the test to run.
2. Hit GitHub API to pull core team from org, if PR author is not in core team; use a https://buildkite.com/docs/pipelines/block-step until a core team member unblocks.

(1) This is out-in-the-open and mirrors other popular open source projects -- like Rust and Swift. But it requires us to maintain a GitHub bot -- though we can do this with a lambda/cloud-function that listens to webhook events on comments.
(2) feels like it's more holistic in the buildkite world, but it's a little more behind-the-scenes and may be opaque to external contributors. Additionally, Brandon's jsonschema-to-dhall library chokes on the block-step schema, so we'd need to update that.

Of these two choices, (1) seems better to me. Feel free to add more ideas in the comments below also.

Changelog:
* Add historical proposals and note the newest is at the bottom
* Change (1) to if PR is from a fork

**Comment by bkase** [comment]
@nholland94 is telling me we can use the github-lambda service instead (since we're open-source it should be free)

**Comment by bkase** [comment]
Dug into a few options for last handful of minutes; rough plan:

1. By default tests do not run on a PR
2. If you add the label, `ci-build-me`, every push of one or more commits will trigger a new build on the tip of the branch (via GitHub Actions)

## Rationale

Tests not only not run on external contributors PRs by default (solving the security issue), but they also don't run for anyone. Why? In progress PRs do not burn CI dollars, and you can turn off CI if you're in the middle of making lots of updates. Since CI spend is a large portion of our infrastructure costs, this is a good way to mitigate some of that.

GitHub Actions: This can be achieved easily, and we don't need to host anything ourselves to do it.

## Not in scope

Cancelling older builds in progress: This is worth doing, but I don't think this existing blocks us in the short term. We can always cull useless jobs by manually cancelling on buildkite's UI.

**Comment by bkase** [comment]
> With the exception of GITHUB_TOKEN, secrets are not passed to the runner when a workflow is triggered from a forked repository.

Grr.. this breaks the flow for external contributors. I'll figure out a workaround

**Comment by bkase** [comment]
This approach is vulnerable to the following attack:

An adversarial contributor first submits a patch, we add the label, then they push code to steal secrets -- since the label is already added CI runs.

I will also include a mechanism for triggering via comment -- it looks like the direct actions approach doesn't quite work since there is no way to expose Buildkite's API token safely.

**Comment by nholland94** [comment]
Commenting each time to allow an external contributor to run CI seems rather unscalable and unacceptable, but so does the vulnerability this pattern entails. How do other organizations deal with this issue? Could we do something like the following?

Bootstrap the buildkite pipelines with a single, truly secure container. This container is will never run if anything involved in running it is touched by an external contributor and has not been explicitly approved by a trusted contributor. When this pipeline runs, it has access to the "true" secrets. From here, we generate temporary secrets with specific access scopes, and forward those to all the future pipelines.

It sounds a little painful to setup, but might fix the problem. But we can literally just generate temp creds with specific API permission scoping. If someone does snoop that secret in one of the future jobs in CI, they will only have snooped a temporary, limited access key that they can't do much with and expires shortly.

**Comment by bkase** [comment]
> Commenting each time to allow an external contributor to run CI seems rather unscalable and unacceptable, but so does the vulnerability this pattern entails. How do other organizations deal with this issue?

Afaik other projects do have comments trigger builds -- at least Swift lang and Rust lang both do this.

Your approach sounds nice, but I don't think it's worth the pain of setting up. For now, would it be acceptable to just have the convention that we only toggle the label on external contributors' PRs rather than leave it on. This way we can manually trigger builds just once.

**Comment by nholland94** [comment]
> Your approach sounds nice, but I don't think it's worth the pain of setting up. For now, would it be acceptable to just have the convention that we only toggle the label on external contributors' PRs rather than leave it on. This way we can manually trigger builds just once.

Toggle sounds easy to mess up. Comment would be better. If we have it as a toggle, it's possible that someone forgets to toggle it, and then you have a serious open vuln.
